File: instalog_ios/ios/Classes/InstalogPlugin.swift
import Flutter
import InstalogIOS
import UIKit

public class InstalogPlugin: NSObject, FlutterPlugin {
    final let TAG = "InstalogPlugin"
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "dev.instalog.flutter/channel", binaryMessenger: registrar.messenger())
        let instance = InstalogPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async { [self] in
            switch call.method {
            case "initialize":
                initialize(call: call, result: result)
            case "show_feedback_modal":
                showFeedbackModal(call: call, result: result)
            case "log":
                logEvent(call: call, result: result)
            case "send_crash":
                sendCrash(call: call, result: result)
            case "identify_user":
                identifyUser(call: call, result: result)
            case "simulate_crash":
                simulateCrash(call: call, result: result)
            default:
                result(FlutterMethodNotImplemented)
                return
            }
        }
    }
    
    func initialize(call: FlutterMethodCall, result: FlutterResult) {
        if let args = call.arguments as? [String: Any] {
            let apiKey = args["api_key"] as? String ?? "no_key"
            let options = args["options"] as? [String: Any] ?? [:]
            
            let instalogOptions = InstalogOptions(dictionary: options)
            Instalog.shared.initialize(key: apiKey, options: instalogOptions)
            result(true)
        } else {
            result(FlutterError.init(code: TAG, message: nil, details: "initialize failed"))
        }
    }
    
    func showFeedbackModal(call: FlutterMethodCall, result: FlutterResult) {
        Instalog.shared.showFeedbackModal()
        result(true)
    }
    
    private func logEvent(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError.init(code: TAG, message: "Missing arguments", details: nil))
            return
        }
        
        guard let event = args["event"] as? String else {
            result(FlutterError.init(code: TAG, message: "Missing event name", details: nil))
            return
        }
        
        let params = args["params"] as? [String: String] ?? [String: String]()
       
        let logModel = InstalogLogModel(event: event, params: params)
        Instalog.shared.logEvent(log: logModel)
        
        do {
            let jsonData = try JSONEncoder().encode(logModel)
            result(true)
        } catch {
            print(error)
            result(
                FlutterError(
                    code: "Instalog.logEvent",
                    message: "Unable to encode event",
                    details: error.localizedDescription
                )
            )
        }
    }
    
    private func sendCrash(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError.init(code: TAG, message: "Missing arguments", details: nil))
            return
        }
        
        let error = args["error"] as? String
        let stack = args["stack"] as? String
        
        Instalog.shared.sendCrash(
            name: error,
            report: stack,
            completion: { sent in
                result(sent)
            }
        )
    }
    
    private func identifyUser(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError.init(code: TAG, message: "Missing arguments", details: nil))
            return
        }
        
        guard let id = args["id"] as? String else {
            result(FlutterError.init(code: TAG, message: "Missing user ID", details: nil))
            return
        }
        
        Instalog.shared.identifyUser(id)
        result(true)
    }
    
    private func simulateCrash(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Instalog.crash.simulateCrash()
        result(true)
    }
}

File: instalog_android/android/src/main/kotlin/dev/instalog/flutter/InstalogPlugin.kt
package dev.instalog.flutter

import android.content.Context
import androidx.annotation.NonNull
import dev.instalog.mobile.*
import dev.instalog.mobile.ui.InstalogAlertData

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import android.app.Activity
import dev.instalog.mobile.models.InstalogLogModel

class InstalogPlugin : FlutterPlugin, MethodCallHandler, InstalogAlertDialogHandler, ActivityAware {
    private val instalog = Instalog.getInstance()

    private lateinit var channel: MethodChannel
    private var context: Context? = null
    private var activity: Activity? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "dev.instalog.flutter/channel")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "get_platform_name" -> {
                result.success("Android")
            }
            "initialize" -> {
                val apiKey = call.argument<String>("api_key") ?: run {
                    result.error("MISSING_API_KEY", "API key is required", null)
                    return
                }
                val options = call.argument<Map<String, Any>>("options") ?: mapOf()
                
                initialize(apiKey, options, result)
            }
            "log" -> {
                val event = call.argument<String>("event") ?: run {
                    result.error("MISSING_EVENT", "Event name is required", null)
                    return
                }
                val params = call.argument<Map<String, Any>>("params") ?: mapOf()
                
                log(event, params, result)
            }
            "identify_user" -> {
                val id = call.argument<String>("id") ?: run {
                    result.error("MISSING_ID", "User ID is required", null)
                    return
                }
                
                identifyUser(id, result)
            }
            "send_crash" -> {
                val error = call.argument<String>("error") ?: run {
                    result.error("MISSING_ERROR", "Error message is required", null)
                    return
                }
                val stack = call.argument<String>("stack") ?: ""
                
                sendCrash(error, stack, result)
            }
            "show_feedback_modal" -> {
                showFeedbackModal(result)
            }
            "simulate_crash" -> {
                simulateCrash(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initialize(apiKey: String, options: Map<String, Any>, result: Result) {
        if (activity == null) {
            result.error("ACTIVITY_NULL", "Current activity is null", null)
            return
        }

        val opt = InstalogOptions(
            isLogEnabled = options["is_log_enabled"] as? Boolean ?: false,
            isLoggerEnabled = options["is_logger_enabled"] as? Boolean ?: false,
            isCrashEnabled = options["is_crash_enabled"] as? Boolean ?: false,
            isFeedbackEnabled = options["is_feedback_enabled"] as? Boolean ?: false
        )

        try {
            instalog.initialize(
                key = apiKey,
                context = activity!!.applicationContext,
                handler = this,
                options = opt
            )
            result.success(true)
        } catch (e: Exception) {
            result.error("INITIALIZE_ERROR", e.message, e)
        }
    }

    private fun log(event: String, params: Map<String, Any>, result: Result) {
        if (activity == null) {
            result.error("ACTIVITY_NULL", "Current activity is null", null)
            return
        }

        try {
            val paramsMap: HashMap<String, String> = params.mapValues { 
                it.value.toString() ?: ""
            } as HashMap<String, String>

            val log = InstalogLogModel(
                event = event,
                params = paramsMap
            )

            instalog.logEvent(activity!!.applicationContext, log)
            result.success(true)
        } catch (e: Exception) {
            result.error("LOG_ERROR", e.message, e)
        }
    }

    private fun identifyUser(id: String, result: Result) {
        try {
            instalog.identifyUser(id)
            result.success(true)
        } catch (e: Exception) {
            result.error("IDENTIFY_USER_ERROR", e.message, e)
        }
    }

    private fun sendCrash(error: String, stack: String, result: Result) {
        if (activity == null) {
            result.error("ACTIVITY_NULL", "Current activity is null", null)
            return
        }

        try {
            instalog.sendCrash(activity!!.applicationContext, error, stack)
            result.success(true)
        } catch (e: Exception) {
            result.error("SEND_CRASH_ERROR", e.message, e)
        }
    }

    private fun showFeedbackModal(result: Result) {
        if (activity == null) {
            result.error("ACTIVITY_NULL", "Current activity is null", null)
            return
        }

        try {
            instalog.showFeedbackModal(activity!!)
            result.success(true)
        } catch (e: Exception) {
            result.error("FEEDBACK_MODAL_ERROR", e.message, e)
        }
    }

    private fun simulateCrash(result: Result) {
        try {
            Instalog.crash.simulateExceptionCrash()
            result.success(true)
        } catch (e: Exception) {
            result.error("SIMULATE_CRASH_ERROR", e.message, e)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    override fun show(data: InstalogAlertData) {
        TODO("Not yet implemented")
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}